drivers/ethernet/phy/phy_tja1103.c

```diff
@@ -128,23 +148,172 @@ static int phy_tja1103_id(const struct device *dev, uint32_t *phy_id)
 	return 0;
 }
 
-static int phy_tja1103_get_link_state(const struct device *dev, struct phy_link_state *state)
+static int update_link_state(const struct device *dev)
 {
 	struct phy_tja1103_data *const data = dev->data;
+	bool link_up;
 	uint16_t val;
 
-	k_sem_take(&data->sem, K_FOREVER);
+	if (phy_tja1103_c45_read(dev, MDIO_MMD_VENDOR_SPECIFIC1, TJA1103_PHY_STATUS, &val) < 0) {
+		return -EIO;
+	}
+
+	link_up = (val & TJA1103_PHY_STATUS_LINK_STAT) != 0;
+
+	/* Let workqueue re-schedule and re-check if the
+	 * link status is unchanged this time
+	 */
+	if (data->state.is_up == link_up) {
+		return -EAGAIN;
+	}
+
+	data->state.is_up = link_up;
 
-	if (phy_tja1103_c45_read(dev, MDIO_MMD_VENDOR_SPECIFIC1, TJA1103_PHY_STATUS, &val) >= 0) {
+	return 0;
+}
+
+static int phy_tja1103_get_link_state(const struct device *dev, struct phy_link_state *state)
+{
+	struct phy_tja1103_data *const data = dev->data;
+	const struct phy_tja1103_config *const cfg = dev->config;
+	k_timeout_t sem_flags;
+	int rc = 0;
+
+	if (k_is_in_isr()) {
+		sem_flags = K_NO_WAIT;
+	} else {
+		sem_flags = K_FOREVER;
+	}
 
-		data->state.is_up = (val & TJA1103_PHY_STATUS_LINK_STAT) != 0;
+	k_sem_take(&data->sem, sem_flags);
 
-		memcpy(state, &data->state, sizeof(struct phy_link_state));
+	/* If Interrupt is configured then the workqueue will not
+	 * update the link state periodically so do it explicitly
+	 */
+	if (cfg->gpio_interrupt.port != NULL) {
+		rc = update_link_state(dev);
 	}
 
+	memcpy(state, &data->state, sizeof(struct phy_link_state));
+
 	k_sem_give(&data->sem);
 
-	return 0;
+	return rc;
+}
+
+static void invoke_link_cb(const struct device *dev)
+{
+	struct phy_tja1103_data *const data = dev->data;
+	struct phy_link_state state;
+
+	if (data->cb == NULL) {
+		return;
+	}
+
+	/* Send callback only on link state change */
+	if (phy_tja1103_get_link_state(dev, &state) != 0) {
+		return;
+	}
+
+	data->cb(dev, &state, data->cb_data);
+}
+
+static void monitor_work_handler(struct k_work *work)
+{
+	struct k_work_delayable *dwork = k_work_delayable_from_work(work);
+	struct phy_tja1103_data *const data =
+		CONTAINER_OF(dwork, struct phy_tja1103_data, monitor_work);
+	const struct device *dev = data->dev;
+	int rc;
+
+	k_sem_take(&data->sem, K_FOREVER);
+
+	rc = update_link_state(dev);
+
+	k_sem_give(&data->sem);
+
+	/* If link state has changed and a callback is set, invoke callback */
+	if (rc == 0) {
+		invoke_link_cb(dev);
+	}
+
+	/* Submit delayed work */
+	k_work_reschedule(&data->monitor_work, K_MSEC(CONFIG_PHY_MONITOR_PERIOD));
+}
+
+static void phy_tja1103_handle_irq(const struct device *port, struct gpio_callback *cb,
```

Comment:
MDIO is a slow blocking protocol. You don't want to do these transfers in irq context since it destroys determinism of the system.

Reply:
Every ISR needs to atleast check the reason of interrupt and try to ack it by accessing the registers of the device that caused it.
Its just that in Ethernet PHY case these device registers are not memory mapped so what's the alternative then ?
Also check out TJA1103 Linux driver, it does the same-
drivers/net/phy/nxp-c45-tja11xx.c (nxp_c45_handle_interrupt)

Reply:
I have done it using an offload thread and a semaphore..  any comments are welcome